reader_at.go

```diff
@@ -0,0 +1,257 @@
+		}
+
+		// the end of the next stream reading backwards is one before the start
+		// of the one we just processed.
+		streamEnd = streamStart - 1
+	}
+
+	return r, nil
+}
+
+// An index carries all the information necessary for reading randomly into a
+// single stream.
+type index struct {
+	blockStartOffset int64
+	streamHeader     streamHeader
+	records          []record
+}
+
+func (i index) compressedBufferedSize() int64 {
+	size := int64(0)
+	for _, r := range i.records {
+		size += r.paddedLen()
+	}
+	return size
+}
+
+func (i index) uncompressedSize() int64 {
+	size := int64(0)
+	for _, r := range i.records {
+		size += r.uncompressedSize
+	}
+	return size
+}
+
+// setupIndexAt takes the offset of the end of a stream, or null bytes following
+// the end of a stream. It builds an index for that stream, adds it to the
+// beginning of the ReaderAt and returns the offset to the beginning of the stream.
+func (r *ReaderAt) setupIndexAt(endOffset int64) (int64, error) {
+	// read backwards past potential null bytes until we find the end of the
+	// footer
+	for endOffset > 0 {
+		probe := make([]byte, 1)
+		n, err := r.xz.ReadAt(probe, endOffset)
+		if err != nil {
+			return 0, err
+		}
+		if n != len(probe) {
+			return 0, fmt.Errorf("read %d bytes", n)
+		}
+		if probe[0] != 0 {
+			break
+		}
+		endOffset--
+	}
+	endOffset++
+
+	footerOffset := endOffset - footerLen
+	f, err := readFooter(newRat(r.xz, footerOffset))
+	if err != nil {
+		return 0, err
+	}
+
+	indexStartOffset := footerOffset - f.indexSize
+
+	// readIndexBody assumes the indicator byte has already been read
+	indexRecs, _, err := readIndexBody(newRat(r.xz, indexStartOffset+1))
+	if err != nil {
+		return 0, err
+	}
+
+	ix := index{
+		records: indexRecs,
+	}
+	ix.blockStartOffset = indexStartOffset - ix.compressedBufferedSize()
+	r.indices = append([]index{ix}, r.indices...)
+
+	sh := streamHeader{}
+	headerStartOffset := ix.blockStartOffset - HeaderLen
+	err = sh.UnmarshalReader(newRat(r.xz, headerStartOffset))
+	if err != nil {
+		return 0, fmt.Errorf("trouble reading stream header at offset %d: %v", headerStartOffset, err)
+	}
+	ix.streamHeader = sh
+
+	xlog.Debugf("xz indices %+v", r.indices)
+
+	return headerStartOffset, nil
+}
+
+func (r *ReaderAt) Size() int64 {
+	total := int64(0)
+	for _, ix := range r.indices {
+		total += ix.uncompressedSize()
+	}
+	return total
+}
+
+func (r *ReaderAt) ReadAt(p []byte, bufferPos int64) (int, error) {
+	lenRequested := len(p)
+
+	indicesPos := int64(0)
+
+	for _, index := range r.indices {
+		blockOffset := index.blockStartOffset
+
+		for _, block := range index.records {
+			if indicesPos <= bufferPos && bufferPos <= indicesPos+block.uncompressedSize {
+				blockStartPos := bufferPos - indicesPos
+				blockEndPos := blockStartPos + int64(len(p))
+				if blockEndPos > block.uncompressedSize {
+					blockEndPos = block.uncompressedSize
+				}
+				blockAmtToRead := blockEndPos - blockStartPos
+
+				r.readBlockAt(
```

Comment:
To do this in parallel this `readBlockAt` could be changed to `errgroup.Go(readBlockAt)`